shared/actions/engine/helper.js

```diff
@@ -0,0 +1,189 @@
+import {call, put, cancelled, fork, join} from 'redux-saga/effects'
+
+import * as SagaTypes from '../../constants/types/saga'
+import * as FluxTypes from '../../constants/types/flux'
+
+// If a sub saga returns bail early, then the rpc will bail early
+const BailEarly = {type: '@@engineRPCCall:bailEarly'}
+const BailedEarly = {type: '@@engineRPCCall:bailedEarly', payload: undefined}
+
+const rpcResult = (args: any) => ({type: '@@engineRPCCall:respondResult', payload: args})
+const rpcError = (args: any) => ({type: '@@engineRPCCall:respondError', payload: args})
+const rpcCancel = (args: any) => ({type: '@@engineRPCCall:respondCancel', payload: args})
+
+const _isResult = ({type} = {}) => type === '@@engineRPCCall:respondResult'
+const _isError = ({type} = {}) => type === '@@engineRPCCall:respondError'
+const _isCancel = ({type} = {}) => type === '@@engineRPCCall:respondCancel'
+
+type Finished = FluxTypes.NoErrorTypedAction<
+  '@@engineRPCCall:finished',
+  {
+    error: ?any,
+    params: ?any,
+  }
+>
+
+const finished = ({error, params}) => ({type: '@@engineRPCCall:finished', payload: {error, params}})
+const isFinished = (a: any) => a.type === '@@engineRPCCall:finished'
+
+type RpcRunResult = Finished | FluxTypes.NoErrorTypedAction<'@@engineRPCCall:bailedEarly', void>
+
+function _sagaWaitingDecorator(rpcNameKey, saga) {
+  return function*(...args: any) {
+    yield put(Creators.waitingForRpc(rpcNameKey, false))
+    yield call(saga, ...args)
+    yield put(Creators.waitingForRpc(rpcNameKey, true))
+  }
+}
+
+// This decorator deals with responding to the rpc
+function _handleRPCDecorator(rpcNameKey, saga) {
+  return function*({params, response}) {
+    const returnVal = yield call(saga, params)
+    const payload = (returnVal || {}).payload
+    if (_isResult(returnVal)) {
+      yield call([response, response.result], payload)
+    } else if (_isCancel(returnVal)) {
+      const engineInst = yield call(engine)
+      yield call([engineInst, engineInst.cancelRPC], response, payload)
+    } else if (_isError(returnVal)) {
+      yield call([response, response.error], payload)
+    } else {
+      throw new Error(`SubSaga for ${rpcNameKey} did not return a response to the rpc!`)
+    }
+  }
+}
+
+function passthroughResponseSaga() {
+  return rpcResult()
+}
+
+class EngineRpcCall {
+  _subSagas: SagaTypes.SagaMap
+  _chanConfig: SagaTypes.ChannelConfig<*>
+  _rpc: Function
+  _rpcNameKey: string // Used for the waiting state and error messages.
+  _request: any
+
+  _engineChannel: EngineChannel
+  _cleanedUp: boolean
+
+  constructor(sagaMap: SagaTypes.SagaMap, rpc: any, rpcNameKey: string, request: any) {
+    this._chanConfig = Saga.singleFixedChannelConfig(Object.keys(sagaMap))
+    this._rpcNameKey = rpcNameKey
+    this._rpc = rpc
+    this._cleanedUp = false
+    this._request = request
+    // $FlowIssue with this
+    this.run = this.run.bind(this) // In case we mess up and forget to do call([ctx, ctx.run])
```

Comment:
I think that's not a valid syntax. There's no fat arrow equivalent of generators yet